### PR TITLE
75 manager sampler add minimum tokens per second accepted

### LINKED ISSUE #75: Manager / Sampler : Add minimum tokens per second accepted
Given a minimum acceptable speed of tokens per second (configured in the Manager) the Sampler should provide the Requested with a time out value that is suitable for each of the generated prompts.

### PR DESCRIPTION
**Pull request by AguirreNicolas**
It was created a TimeoutHandler to assing timeouts.
For LLMs it depends of 3 variables:

- Prefill Time a.k.a Time-To-First-Token (TTFT)
- Decode Time a.k.a Time-Per-Output-Token (TPOT)
- Queue Time

**Prefill time (TTFT)**

It was modeled as a quadratic functions [1]. Besides  (0, 0), two more points are nedded. This two point acts as a SLA. For insntance, this points are (8192,2) and (32768, 10).

image

**Decode time (TPOT)**

The decode time is modeled as a linear function with respect to the average silent reading time in English. According to the reference [2], the average silent reading speed is 238 words per minute. Given that there are approximately 0.75 tokens per word, the TPOT can be calculated as follows:

$`TPOT ~(ms/tok) = \frac{1000 ms}{1seg *~\frac{238 ~words}{min}*\frac{ 0.75 ~tok}{word}*\frac{1min}{60seg} } = \frac{1000ms}{2.975tok}  = 336  \frac{ms}{tok}`$

**Queue time**

The queue time has been defined as 30 seconds.

Close #75 

[1] https://arxiv.org/abs/2407.07000
[2] https://www.sciencedirect.com/science/article/abs/pii/S0749596X19300786

Close #75

### CODE REVIEW

File: docker-compose/dev/apps/config/sampler.json [hunk 1]
```diff
@@ -7,5 +7,23 @@
     "port": 7233,
     "namespace": "pocket-ml-testbench",
     "task_queue": "sampler"
-  }
+  },
+  "timeouts": {
```
Comment (RawthiL) [inline docker-compose/dev/apps/config/sampler.json:11]: "timeouts" should be nested inside a "LLM" category, it should not be a base configuration of the sampler app.

File: docker-compose/dev/apps/config/sampler.json [hunk 2]
```diff
@@ -7,5 +7,23 @@
...
+    "random": {
+      "ttft": {
+        "x": [
```
Comment (RawthiL) [inline docker-compose/dev/apps/config/sampler.json:14]: "x" and "y" offer no information on what they actually are, use more descriptive labels

File: docker-compose/morse-poc/apps_configs/sampler.json [hunk 3]
```diff
@@ -12,5 +12,87 @@
     "max_concurrent_workflow_tasks": 4,
     "max_concurrent_workflow_task_polls": 4,
     "max_concurrent_activity_task_polls": 4
-  }
+  },
+  "timeouts": {
```
Comment (RawthiL) [inline docker-compose/morse-poc/apps_configs/sampler.json:16]: same as docker-compose/dev

File: apps/python/sampler/activities/lmeh/sample.py [hunk 4]
```diff
@@ -19,7 +23,10 @@ async def lmeh_sample(args: PocketNetworkTaskRequest) -> bool:
     eval_logger = get_app_logger("sample")
     config = get_app_config()["config"]
     wf_id = activity.info().workflow_id
-
+    timeouts = LLMTimeouts(**config["timeouts"][args.requester_args.service])
```
Comment (RawthiL) [inline apps/python/sampler/activities/lmeh/sample.py:26]: Handle the case when no timeouts config is provided.

File: packages/python/protocol/protocol.py [hunk 5]
```diff
@@ -310,3 +311,62 @@ class PocketNetworkMongoDBConfig(BaseModel):
...
+class TimeoutHandler(BaseModel):
+    service: str
+    timeouts: LLMTimeouts
+
+    # Define the functions for each chain
+    def timeouts_a100(self, prefill: int, decode: int) -> float:
+        x = self.timeouts.ttft.x
+        y = self.timeouts.ttft.y
+        queue = self.timeouts.queue
+        z = np.polyfit(x, y, 2)
```
Comment (RawthiL) [inline packages/python/protocol/protocol.py:341]: you are calling polyfit on each timeout calculation request?

Also, handle when this polyfit fails (configuration can be wrong)

**Review by RawthiL — CHANGES_REQUESTED** [review]
some minor changes requiered

File: packages/python/protocol/protocol.py (hunk 5 above)
Comment (AguirreNicolas) [reply packages/python/protocol/protocol.py:341]: If the timeout correspond to a 'llm', now it only fit one time.

File: docker-compose/dev/apps/config/sampler.json (hunk 1 above)
Comment (AguirreNicolas) [reply docker-compose/dev/apps/config/sampler.json:11]: But the only information accesed is the service. I added a 'type' attr. Now it only take the value of 'llm'. In the future itg can vary.

File: docker-compose/dev/apps/config/sampler.json (hunk 2 above)
Comment (AguirreNicolas) [reply docker-compose/dev/apps/config/sampler.json:14]: Fixed, named as "prompt_lenght" and "sla_time" (service level aggrement).

**Review by RawthiL — APPROVED** [review]

**Pull request merged**